web/modules/custom/joinup_user/joinup_user.behat.inc

```diff
@@ -58,6 +58,41 @@ class JoinupUserSubContext extends DrupalSubContextBase implements DrupalSubCont
     $this->visitPath(user_pass_reset_url($user) . '/login');
   }
 
+  /**
+   * Clicks an html link in the last email matching its title.
+   *
+   * @param string $title
+   *   The link title.
+   * @param string $user
+   *   The user that the email was sent to.
+   *
+   * @Given /^I click the "(?P<title>[^"]+)" link from the email sent to "(?P<user>[^"]+)"$/
```

Comment:
We can make this easier to read by replacing the regular expressions with the string substitutions that were introduced in Behat 3.0:

```
@Given I click the :title link from the email sent to :user
```

Reply:
I know... But there seems to be a compatibility problem with my PHP Storm and when I leave it only to `@Given I click the :title link from the email sent to :user` and I start typing, I only get the `@Given I click the <param>" and stops there. And I do not see the full step definition..

Reply:
I researched this and it turns out that the PHPStorm plugin (ref. https://plugins.jetbrains.com/plugin/9164-gherkin) is following the reference implementation of Gherkin that is supplied by the Cucumber project.

The reference implementation supports "Cucumber expressions" which are in this format:

```
@Given I click the {string} link in the email sent to {string}
```

The `:title` syntax for variable substitution that the "Behat Gherkin" implementation supports is not included in the "Cucumber Gherkin" specification (ref. https://cucumber.io/docs/cucumber/step-definitions/ and https://cucumber.io/docs/cucumber/cucumber-expressions/).

The regex variant is supported by both. Unfortunately it seems this is not working as expected in PHPStorm either: https://youtrack.jetbrains.com/issue/WI-25135

Also the regex variant is a nightmare to review. The fact that this is copied from some other part of the code does not make me trust it :)

I am quite fluent with regular expressions but I had to google the `?P` part (it turns out to be a _symbolic group match_ ref. https://docs.python.org/3/library/re.html#index-17). The `:title` syntax on the other hand is very clear and easy to review.

Now I think that we should not make decisions based on whether or not something works in PHPStorm. We do not require people to use this proprietary software for working on Joinup. There are 2 possibilities: either Jetbrains can include support for the "Behat Gherkin" syntax in their proprietary Behat plugin, or the "Behat Gherkin" implementation can add support for the "Cucumber Expressions". Both of these solutions would make it work out of the box for everyone. I will make a feature request to support this in Behat.

Now, on the other hand now that I have fully reviewed the regular expression with the symbolic group match I know it is correct, so I have no reason to reject it any more. It can stay :)

Reply:
I did some more research and the "Cucumber expressions" syntax results in the same problems as the "Behat syntax". Only the "regex syntax" works in PHPStorm. It has been reported in their bug tracker: https://youtrack.jetbrains.com/issue/WI-49121

Reply:
According to the above, I will leave it with the `:parameter` syntax and hope it is fixed upstream soon.